src/libraries/System.Private.CoreLib/src/System/Numerics/Decimal128.cs

```diff
@@ -0,0 +1,268 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers.Text;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Text;
+
+namespace System.Numerics
+{
+    public readonly struct Decimal128
+        : IComparable,
+          IComparable<Decimal128>,
+          IEquatable<Decimal128>,
+          ISpanParsable<Decimal128>,
+          IDecimalIeee754ParseAndFormatInfo<Decimal128>,
+          IDecimalIeee754ConstructorInfo<Decimal128, Int128, UInt128>,
+          IDecimalIeee754UnpackInfo<Decimal128, Int128, UInt128>,
+          IDecimalIeee754TryParseInfo<Decimal128, Int128>
+    {
+#if BIGENDIAN
+        internal readonly ulong _upper;
+        internal readonly ulong _lower;
+#else
+        internal readonly ulong _lower;
+        internal readonly ulong _upper;
+#endif
+
+        private const int MaxDecimalExponent = 6111;
+        private const int MinDecimalExponent = -6176;
+        private const int NumberDigitsPrecision = 34;
+        private const int Bias = 6176;
+        private const int NumberBitsExponent = 14;
+        private static readonly UInt128 PositiveInfinityValue = new UInt128(upper: 0x7800000000000000, lower: 0);
+        private static readonly UInt128 NegativeInfinityValue = new UInt128(upper: 0xf800000000000000, lower: 0);
+        private static readonly UInt128 ZeroValue = new UInt128(0, 0);
+        private static readonly Int128 MaxSignificand = new Int128(upper: 542101086242752, lower: 4003012203950112767); // 9_999_999_999_999_999_999_999_999_999_999_999;
+        private static readonly UInt128 SignMask = new UInt128(0x8000_0000_0000_0000, 0);
+        private static readonly UInt128 G0G1Mask = new UInt128(0x6000_0000_0000_0000, 0);
+        private static readonly UInt128 MostSignificantBitOfSignificandMask = new UInt128(0x0002_0000_0000_0000, 0);
```

Comment:
Many of these would be better as properties, as it allows the JIT to inline the constructed value and more trivially see the fields are constant.

There is some JIT support for observing such constants through `static readonly`, but it's often not as good.